scripts/Interface/reduction_gui/reduction/toftof/toftof_reduction.py

```diff
@@ -0,0 +1,484 @@
+#pylint: disable = too-many-instance-attributes, too-many-locals, too-many-branches
+#pylint: disable = attribute-defined-outside-init
+#pylint: disable = invalid-name
+#pylint: disable = W0622
+"""
+TOFTOF reduction workflow gui.
+"""
+import xml.dom.minidom
+
+from PyQt4.QtCore import *
+from PyQt4.QtGui  import *
+
+from reduction_gui.reduction.scripter import BaseScriptElement, BaseReductionScripter
+
+#-------------------------------------------------------------------------------
+
+class TOFTOFScriptElement(BaseScriptElement):
+
+    # normalisation
+    NORM_NONE       = 0
+    NORM_MONITOR    = 1
+    NORM_TIME       = 2
+
+    # TOF correction
+    CORR_TOF_NONE   = 0
+    CORR_TOF_VAN    = 1
+    CORR_TOF_SAMPLE = 2
+
+    # default values
+    DEF_prefix     = 'ws'
+    DEF_ecFactor   = 1.0
+
+    DEF_binEon     = True
+    DEF_binEstart  = 0.0
+    DEF_binEstep   = 0.0
+    DEF_binEend    = 0.0
+
+    DEF_binQon     = True
+    DEF_binQstart  = 0.0
+    DEF_binQstep   = 0.0
+    DEF_binQend    = 0.0
+
+    DEF_subECVan   = False
+    DEF_normalise  = NORM_NONE
+    DEF_correctTof = CORR_TOF_NONE
+
+    XML_TAG = 'TOFTOFReduction'
+
+    def reset(self):
+        self.facility_name   = ''
+        self.instrument_name = ''
+
+        # prefix of (some) workspace names
+        self.prefix   = self.DEF_prefix
+
+        # data files are here
+        self.dataDir  = ''
+
+        # vanadium runs & comment
+        self.vanRuns  = ''
+        self.vanCmnt  = ''
+
+        # empty can runs, comment, and factor
+        self.ecRuns   = ''
+        self.ecFactor = self.DEF_ecFactor
+
+        # data runs: [(runs,comment), ...]
+        self.dataRuns = []
+
+        # additional parameters
+        self.binEon        = self.DEF_binEon
+        self.binEstart     = self.DEF_binEstart
+        self.binEstep      = self.DEF_binEstep
+        self.binEend       = self.DEF_binEend
+
+        self.binQon        = self.DEF_binQon
+        self.binQstart     = self.DEF_binQstart
+        self.binQstep      = self.DEF_binQstep
+        self.binQend       = self.DEF_binQend
+
+        self.maskDetectors = ''
+
+        # options
+        self.subtractECVan = self.DEF_subECVan
+        self.normalise     = self.DEF_normalise
+        self.correctTof    = self.DEF_correctTof
+
+    def to_xml(self):
+        res = ['']
+
+        def put(tag, val):
+            res[0] += '  <%s>%s</%s>\n' % (tag, str(val), tag)
```

Comment:
To be future-proof, the new-style string formatting `"{}".format(var)` should be preferred.